src/lib/Zikula/Core/Theme/AssetBag.php

```diff
@@ -19,18 +39,22 @@ public function __construct(array $assets = array())
      */
     public function add($asset)
     {
+        // ensure value is an array
         if (!is_array($asset)) {
-            $asset = array($asset);
+            $asset = [$asset => self::DEFAULT_WEIGHT];
         }
-        $this->assets = array_merge($this->assets, $asset);
-        $this->assets = array_unique($this->assets);
+        foreach ($asset as $source => $weight) {
+            if ((isset($this->assets[$source]) && $this->assets[$source] > $weight) || !isset($this->assets[$source])) {
+                // keep original weight if lighter. set if not set already.
+                $this->assets[$source] = $weight;
+            }
+        }
+        asort($this->assets); // put array in order by weight
```

Comment:
smarty's `{pageaddvar}` uses PageUtil. The (legacy) values are then pulled into the new theme engine [here](https://github.com/zikula/core/pull/2606/files#diff-618272da4726952c6a95d46abcdba26cL343) and the order of those is maintained as is (note the `$i` counter).